gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+ * ephemeral port.
+ *
+ * <p>While the mock attempts to simulate the service, there are some differences in the behaviour.
+ * The mock will accept any project ID and never returns a notFound or another error because of
+ * project ID. It assumes that all project IDs exist and that the user has all the necessary
+ * privileges to manipulate any project. Similarly, the local simulation does not work with any
+ * verification of domain name ownership. Any request for creating a managed zone will be approved.
+ * The mock does not track quota and will allow the user to exceed it. The mock provides only basic
+ * validation of the DNS data for records of type A and AAAA. It does not validate any other record
+ * types.
+ */
+public class LocalDnsHelper {
+
+  private final ConcurrentSkipListMap<String, ProjectContainer> projects
+      = new ConcurrentSkipListMap<>();
+  private static final URI BASE_CONTEXT;
+  private static final Logger log = Logger.getLogger(LocalDnsHelper.class.getName());
+  private static final JsonFactory jsonFactory = new JacksonFactory();
+  private static final Random ID_GENERATOR = new Random();
+  private static final String VERSION = "v1";
+  private static final String CONTEXT = "/dns/" + VERSION + "/projects";
+  private static final Set<String> SUPPORTED_COMPRESSION_ENCODINGS =
+      ImmutableSet.of("gzip", "x-gzip");
+  private static final List<String> TYPES = ImmutableList.of("A", "AAAA", "CNAME", "MX", "NAPTR",
+      "NS", "PTR", "SOA", "SPF", "SRV", "TXT");
+
+  static {
+    try {
+      BASE_CONTEXT = new URI(CONTEXT);
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(
+          "Could not initialize LocalDnsHelper due to URISyntaxException.", e);
+    }
+  }
+
+  private long delayChange;
+  private final HttpServer server;
+  private final int port;
+
+  /**
+   * For matching URLs to operations.
+   */
+  private enum CallRegex {
+    CHANGE_CREATE("POST", CONTEXT + "/[^/]+/managedZones/[^/]+/changes"),
+    CHANGE_GET("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/changes/[^/]+"),
+    CHANGE_LIST("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/changes"),
+    ZONE_CREATE("POST", CONTEXT + "/[^/]+/managedZones"),
+    ZONE_DELETE("DELETE", CONTEXT + "/[^/]+/managedZones/[^/]+"),
+    ZONE_GET("GET", CONTEXT + "/[^/]+/managedZones/[^/]+"),
+    ZONE_LIST("GET", CONTEXT + "/[^/]+/managedZones"),
+    PROJECT_GET("GET", CONTEXT + "/[^/]+"),
+    RECORD_LIST("GET", CONTEXT + "/[^/]+/managedZones/[^/]+/rrsets");
+
+    private String method;
+    private String pathRegex;
+
+    CallRegex(String method, String pathRegex) {
+      this.pathRegex = pathRegex;
+      this.method = method;
+    }
+  }
+
+  /**
+   * Wraps DNS data by adding a timestamp and id which is used for paging and listing.
+   */
+  static class RrsetWrapper {
+    static final Function<ResourceRecordSet, RrsetWrapper> WRAP_FUNCTION =
+        new Function<ResourceRecordSet, RrsetWrapper>() {
+          @Nullable
```

Comment:
So far, we haven't used this annotation in gcloud-java.  I think we should either use it consistently across the library or not at all.